Feat/ping and heartbeat

Korfeusz opened the pull request:
See issue: #7  
Added a ping and a monitor method to Machine as well as a new example.
Tried to keep it as similar to the keygen-go implementation, except for:

- it does not panic on a failed heartbeat, but  transmits the error through an async channel to be handled by the caller for more flexibility. 
- also it does not read the interval from the Machine struct, because it can be None, the caller can decide on a good default. 

Also I added the futures-timer crate as a dependency so that there is an async-runtime agnostic delay.

ahonn commented:
Everything looks great! I will test it tomorrow and merge it as soon as possible.

ahonn commented on examples/start_heartbeat.rs:
```diff
@@ -0,0 +1,45 @@
+use std::{
```
Missing `std::env` and `dotenv::dotenv` caused the error for `cargo run --example start_heartbeat`. 
The following code needs to be added.   

```rust
use dotenv::dotenv;
use std::env;

#[tokio::main]
async fn main() -> Result<(), Error> {
    dotenv().ok();
    ...
}
```

ahonn commented on examples/start_heartbeat.rs:
```diff
@@ -0,0 +1,45 @@
+    let fingerprint = machine_uid::get().unwrap_or("".into());
+    if let Ok(license) = keygen_rs::validate(&[fingerprint.clone()], &[]).await {
+        let machine = license.machine(&fingerprint).await?;
+        let interval = Duration::from_secs(machine.heartbeat_duration.unwrap_or(570) as u64);
```
The default `heartbeat_duration` is 60s.

see https://keygen.sh/docs/api/policies/#policies-object-attrs-heartbeatDuration

Korfeusz replied:
The link says that 60s is the minimum heartbeat duration, I would be afraid to set something to the absolute minimum. Also the dropdown on app.keygen.sh when setting up a new policy does not even have an option to go under 2 minutes. 
Also here: https://keygen.sh/docs/api/machines/#machines-actions-ping it says that the default heartbeat ping window is 10 minutes. I have set it to 10 minutes - 30 seconds, similar to the go reference.

ahonn commented on src/machine.rs:
```diff
@@ -110,4 +114,45 @@ impl Machine {
+
+
+    pub fn monitor(self: Arc<Self>, heartbeat_interval: Duration, tx: Option<Sender<Error>>) -> BoxFuture<'static, ()> {
+        
```
Reference https://github.com/keygen-sh/keygen-go/blob/master/machine.go#L147, I think we should immediately call the `ping` method to send a heartbeat at first.

```rust
pub fn monitor(
    self: Arc<Self>,
    heartbeat_interval: Duration,
    tx: Option<Sender<Error>>,
) -> BoxFuture<'static, ()> {
    async move {
        if let Err(e) = self.ping().await {
            match &tx {
                Some(tx) => {
                    tx.send(e).unwrap();
                }
                _ => {
                    log::error!("Heartbeat monitor failed, aborting.");
                }
            };
            return;
        }

        ...
  }

```

ahonn commented on src/machine.rs:
```diff
@@ -110,4 +114,45 @@ impl Machine {
+        
+        async move {
+            let mut interval_stream = futures::stream::unfold((), move |_| {
+                let delay = futures_timer::Delay::new(heartbeat_interval);
```
Suppose we set the heartbeat interval according to the policy `heartbeatDuration`. In that case, it may not be sent on time due to network issues.

Referring to the code from https://github.com/keygen-sh/keygen-go/blob/master/machine.go#L152, I also recommend using the same way to subtract 30 seconds from the `heartbeat_interval.`

```rust
let delay = futures_timer::Delay::new(heartbeat_interval - Duration::from_secs(30));
```

Korfeusz replied:
The heartbeat duration is a parameter, to the method and so I wanted the final user to be able to set the parameter to whatever they want. If we then subtracted 30s from it it could be surprising behavior. In go the duration is not a parameter as it is being read from the configuration, which we can not do as the config has an Option on the HeartbeatDuration. Which would mean we have to decide on a default value or throw an error when it is not available and I would prefer to leave that decision to the user of the library and their preferences.

ahonn replied:
That makes sense, but I think we should reflect a bit on the example. 

Changing `let interval = Duration::from_secs(machine.heartbeat_duration.unwrap_or(570) as u64);` to `let interval = Duration::from_secs(machine.heartbeat_duration.unwrap_or(600) - 30 as u64);` would be better.

ahonn commented:
I have left some comments on the code. Everything else looks good. 
Thank you for your contribution. Once you address those issues, I'll be happy to merge them.

Korfeusz commented:
Thank you for the review! I have implemented some of the comments and added a reply where I think another decision would also be possible and given my reasoning.

ahonn commented:
Thank you for your contribution. I just merged this PR.
I made some updates to enhance it, and now it supports the cancel monitor: https://github.com/ahonn/keygen-rs/commit/05b493e80ab31c95606d40aa26a8154dc69005e5

Korfeusz commented:
Thank you for merging and the enhancement!
